apisix/http/route.lua

```diff
@@ -111,7 +111,8 @@ function _M.match_uri(uri_router, api_ctx)
     match_opts.vars = api_ctx.var
     match_opts.matched = core.tablepool.fetch("matched_route_record", 0, 4)
 
-    local ok = uri_router:dispatch(api_ctx.var.uri, match_opts, api_ctx, match_opts)
+    local encoded_uri = core.utils.uri_safe_encode(api_ctx.var.uri)
```

Comment:
I have two concerns:
1. In the `radixtree_uri` matching mode, the `core.utils.uri_safe_encode` function will also be run, but it actually has no effect.
2. The probability that the uri parameter contains non-ASCII characters is low. Is there a more efficient way to solve this problem?

Reply:
only encode the uri if router is `radixtree_uri_with_parameter`? or it's better the `lua-resty-radixtree` do the uri encode, while it has the `global_opts.no_param_match` and claim the `res[i] = [=[([\w\-_;:@&=!',\%\$\.\+\*\(\)]+)]=]`? 

and we need a func to check if the uri contains any non-ASCII character

Reply:
@wklken what happens if uri contains non-ascii characters?

Reply:
We can't modify like this, let me show you an example
```
=== TEST 1: set route(id: 1)
--- config
    location /t {
        content_by_lua_block {
            local t = require("lib.test_admin").test
            local code, body = t('/apisix/admin/routes/1',
                 ngx.HTTP_PUT,
                 [[{
                        "methods": ["GET"],
                        "upstream": {
                            "nodes": {
                                "127.0.0.1:1980": 1
                            },
                            "type": "roundrobin"
                        },
                        "uris": ["/hello","/hello1","/中文"]
                }]]
                )

            if code >= 300 then
                ngx.status = code
            end
            ngx.say(body)
        }
    }
--- request
GET /t
--- response_body
passed



=== TEST 2: hit routes1
--- request
GET /中文
--- response_body
hello world
```
The non-ascii uri will hit successfully before your change, but fails after your change

Reply:
agree, the lua-resty-radixtree library should fix it